File: backend/routes/careplan.js
const express = require("express");
const router = express.Router();

const auth = require("../middleware/auth");
const role = require("../middleware/role");
const CarePlan = require("../models/careplan");
const Doctor = require("../models/doctor");
const Patient = require("../models/patient");
const User = require("../models/user"); // Fallback for old data

router.post("/", auth, role("doctor"), async (req, res) => {
  const plan = await CarePlan.create({
    appointmentId: req.body.appointmentId,
    patientId: req.body.patientId,
    doctorId: req.user.id,
    medications: req.body.medications,
    exercises: req.body.exercises,
    instructions: req.body.instructions,
    warningSigns: req.body.warningSigns,
    pdfUrl: req.body.pdfUrl,
  });
  res.status(201).json(plan);
});

router.get("/:patientId", auth, async (req, res) => {
  try {
    const plans = await CarePlan.find({ patientId: req.params.patientId });
    console.log(`Found ${plans.length} care plans for patient ${req.params.patientId}`);

    // Populate doctor and patient names
    const plansWithNames = await Promise.all(
      plans.map(async (plan) => {
        console.log(`Processing plan ${plan._id}, doctorId: ${plan.doctorId}`);
        
        // Try new Doctor model first
        let doctor = await Doctor.findById(plan.doctorId).select("name");
        
        // Fallback to old User model if not found
        if (!doctor) {
          console.log(`Doctor not found in Doctor model, checking User model...`);
          doctor = await User.findById(plan.doctorId).select("name");
        }
        
        // Try new Patient model first
        let patient = await Patient.findById(plan.patientId).select("name");
        
        // Fallback to old User model if not found
        if (!patient) {
          console.log(`Patient not found in Patient model, checking User model...`);
          patient = await User.findById(plan.patientId).select("name");
        }
        
        console.log(`Doctor found: ${doctor?.name}, Patient found: ${patient?.name}`);

        return {
          ...plan.toObject(),
          doctorName: doctor?.name || "Unknown Doctor",
          patientName: patient?.name || "Unknown Patient",
        };
      })
    );

    console.log(`Returning ${plansWithNames.length} plans with names`);
    res.json(plansWithNames);
  } catch (error) {
    console.error('Error in careplan GET route:', error);
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: application/lib/models/care_plan.dart
class CarePlan {
  final String id;
  final String patientId;
  final String doctorId;
  final String? appointmentId;
  final DateTime createdAt;
  final List<Medication> medications;
  final List<Exercise> exercises;
  final String instructions;
  final String warningSigns;
  final String? pdfUrl;
  final String? doctorName;
  final String? patientName;

  CarePlan({
    required this.id,
    required this.patientId,
    required this.doctorId,
    this.appointmentId,
    required this.createdAt,
    required this.medications,
    required this.exercises,
    required this.instructions,
    required this.warningSigns,
    this.pdfUrl,
    this.doctorName,
    this.patientName,
  });

  factory CarePlan.fromJson(Map<String, dynamic> json) {
    return CarePlan(
      id: json['_id'] ?? json['id'],
      patientId: json['patientId'],
      doctorId: json['doctorId'],
      appointmentId: json['appointmentId'],
      createdAt: DateTime.parse(json['createdAt']),
      medications:
          (json['medications'] as List)
              .map((m) => Medication.fromJson(m))
              .toList(),
      exercises:
          (json['exercises'] as List).map((e) => Exercise.fromJson(e)).toList(),
      instructions: json['instructions'] ?? '',
      warningSigns: json['warningSigns'] ?? '',
      pdfUrl: json['pdfUrl'],
      doctorName: json['doctorName'],
      patientName: json['patientName'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'patientId': patientId,
      'doctorId': doctorId,
      'appointmentId': appointmentId,
      'createdAt': createdAt.toIso8601String(),
      'medications': medications.map((m) => m.toJson()).toList(),
      'exercises': exercises.map((e) => e.toJson()).toList(),
      'instructions': instructions,
      'warningSigns': warningSigns,
      'pdfUrl': pdfUrl,
      'doctorName': doctorName,
      'patientName': patientName,
    };
  }
}

class Medication {
  final String name;
  final String dosage;
  final String frequency;
  final String? duration;

  Medication({
    required this.name,
    required this.dosage,
    required this.frequency,
    this.duration,
  });

  factory Medication.fromJson(Map<String, dynamic> json) {
    return Medication(
      name: json['name'],
      dosage: json['dosage'],
      frequency: json['frequency'],
      duration: json['duration'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'dosage': dosage,
      'frequency': frequency,
      'duration': duration,
    };
  }
}

class Exercise {
  final String name;
  final String duration;
  final String frequency;

  Exercise({
    required this.name,
    required this.duration,
    required this.frequency,
  });

  factory Exercise.fromJson(Map<String, dynamic> json) {
    return Exercise(
      name: json['name'],
      duration: json['duration'],
      frequency: json['frequency'] ?? 'Daily',
    );
  }

  Map<String, dynamic> toJson() {
    return {'name': name, 'duration': duration, 'frequency': frequency};
  }
}
